### PR TITLE
Make it cold

### PR DESCRIPTION
**Pull request by rsicarelli**
Instead of creating the `Single.create()`, try to use `fromCallable{}` instead. This is a hot vs cold and you can find more details here: https://medium.com/@elye.project/rxjava-2-understanding-hot-vs-cold-with-just-vs-fromcallable-3c463f9f68c9

Also, I added a line of code that will break your unit test :party_parrot:. Could you identify why?

some tips: 
https://medium.com/@fabioCollini/testing-asynchronous-rxjava-code-using-mockito-8ad831a16877
https://medium.com/@PaulinaSadowska/writing-unit-tests-on-asynchronous-events-with-rxjava-and-rxkotlin-1616a27f69aa

### CODE REVIEW

**Comment by matiasdelbel** [comment]
So, related to the `Single.create{ }` vs `Single. fromCallable{ }` the difference is that for each added subscriber, `create{ }` will return the same value because the lambda is evaluated immediately and only once (the Observable produced the data = cold observable), but `fromCallable { }` will evaluate only the expression when a subscriber is added (an external function produce the data = hot observable). Is it right?

Is there a way to mix the two behaviors? I mean, only evaluate the conditions once, when the first subscriber is added, and then return the cached value for all the subscribers? I think that would be the better approach for this problem because `readJsonFromAsset()` is an expensive operation. It will be nice to add the result in a cache when the first subscriber is added and not to read the file each time a new no first subscriber is added. What do you think?

```
    private val jsonFromAssets = Single.fromCallable {
        val jsonStream = assetsManager.open(FILE_NAME)
        jsonStream.bufferedReader().use { it.readText() }
    }

    fun obtainAll() = jsonFromAssets
        .subscribeOn(Schedulers.io()) 
        .map { convertToDto(it) }
```

**Comment by matiasdelbel** [comment]
Regarding the broken test, the line you've added `.subscribeOn(Schedulers.io())` specify that `readJsonFromAsset()` run in another thread. This causes that `readJsonFromAsset()` may run after assertions in the test.

My doubt is which is the better approach for this situation:
1- Declare the method as following
```
fun obtainAll() = Single.fromCallable { readJsonFromAsset() }
        .map { convertToDto(it) }
```
And then the consumer decides where to execute it:
```
val dataSource = AssetsDataSource(assetsManager, jsonParser)
dataSource
            .obtainAll()
            .subscribeOn(Schedulers.io())
            .subscribe { result, error ->
                
            }
```

2- Pass the scheduler to the data source
```
internal class AssetsDataSource constructor(
    private val assetsManager: AssetManager,
    private val jsonParser: Gson,
    private val scheduler: Scheduler
) {

    fun obtainAll() = Single.fromCallable { readJsonFromAsset() }
        .subscribeOn(scheduler)
        .map { convertToDto(it) }
```
When we work with coroutines, we pass the dispatcher this way, and I think they are similar concepts.

**Comment by matiasdelbel** [comment]
I've updated the PR fixing the test...

**Comment by rsicarelli** [comment]
> (...) Is it right?

yes! Score 😆 

> 
> Is there a way to mix the two behaviors? I mean, only evaluate the conditions once, when the first subscriber is added, and then return the cached value for all the subscribers? 

I believe yes, I need to discuss the use case further. Maybe Rx already has some magic for us, check http://reactivex.io/documentation/operators/from.html. If not possible using cold/hot streams, we could achieve this in different ways for sure. But advanced stuff!

**Comment by rsicarelli** [comment]
> Regarding the broken test, the line you've added `.subscribeOn(Schedulers.io())` specify that `readJsonFromAsset()` run in another thread. This causes that `readJsonFromAsset()` may run after assertions in the test.
> 
> A better approach for this situation is to pass the scheduler to the data source
> 
> ```
> internal class AssetsDataSource constructor(
>     private val assetsManager: AssetManager,
>     private val jsonParser: Gson,
>     private val scheduler: Scheduler
> ) {
> 
>     fun obtainAll() = Single.fromCallable { readJsonFromAsset() }
>         .subscribeOn(scheduler)
>         .map { convertToDto(it) }
> ```
> 
> When we work with coroutines, we pass the dispatcher this way, and I think they are similar concepts.

Score! I rather prefer to use RxPlugins to handle this in the test level to avoid adding the Schedulers on every constructor (purely boiler-plate avoidance) like 

```
public class TrampolineSchedulerRule implements TestRule {

  @Override
  public Statement apply(final Statement base, Description d) {
    return new Statement() {
      @Override
      public void evaluate() throws Throwable {
        RxJavaPlugins.setIoSchedulerHandler(
            scheduler -> Schedulers.trampoline());
        RxJavaPlugins.setComputationSchedulerHandler(
            scheduler -> Schedulers.trampoline());
        RxJavaPlugins.setNewThreadSchedulerHandler(
            scheduler -> Schedulers.trampoline());
        RxAndroidPlugins.setInitMainThreadSchedulerHandler(
            scheduler -> Schedulers.trampoline());

        try {
          base.evaluate();
        } finally {
          RxJavaPlugins.reset();
          RxAndroidPlugins.reset();
        }
      }
    };
  }
}
```

File: heritage/gateway/src/test/kotlin/com/delbel/heritage/gateway/source/AssetsDataSourceTest.kt [hunk 1]
```diff
@@ -21,24 +18,18 @@ class AssetsDataSourceTest {
...
         }
-        val dataSource = AssetsDataSource(assetManager, jsonParser = Gson())
+        val dataSource = AssetsDataSource(
+            assetsManager = assetManager,
+            jsonParser = Gson(),
+            scheduler = Schedulers.trampoline()
+        )
 
-        dataSource.obtainAll().subscribe(testObserver)
+        val testObserver = dataSource.obtainAll().test()
```
Comment (rsicarelli) [inline heritage/gateway/src/test/kotlin/com/delbel/heritage/gateway/source/AssetsDataSourceTest.kt]: nice one! Good practice to retrieve the `TestObservable` like this

**Comment by rsicarelli** [comment]
LGTM, you can merge if you want

**Comment by matiasdelbel** [comment]
I like the approach of defining a `TestRule` to switch the test schedulers. The only thing I don't agree completely with is that you are externally changing the behavior of the class that you are testing. For new team members could be difficult to understand what is happening.

I've followed the other path, and I inject the `schedulers` (next PR). It is like you said, a lot of boilerplate code. I had to create a common module, so I can provide the schedulers with Dagger and annotate the injected field with the expected scheduler.

**Pull request merged**